cps/kni.c

```diff
@@ -629,58 +799,76 @@ data_ipv6_attr_cb(const struct nlattr *attr, void *data)
 }
 
 static int
-rd_modroute(const struct nlmsghdr *req, void *arg)
+rd_modroute(const struct nlmsghdr *req)
 {
 	struct nlattr *tb[__RTA_MAX] = {};
 	struct rtmsg *rm = mnl_nlmsg_get_payload(req);
-	struct route_update *update = arg;
+	struct route_update update;
+	struct cps_config *cps_conf = get_cps_conf();
 
-	update->valid = false;
+	if (unlikely(cps_conf->gk == NULL)) {
+		/*
+		 * Grantor only runs CPS for ECMP support and
+		 * shouldn't be receiving route updates.
+		 */
+		CPS_LOG(WARNING,
+			"The system is running as Grantor, and there shouldn't be any rtnetlink message processed under this configuration while receiving route update messages\n");
+		return MNL_CB_OK;
+	}
 
 	CPS_LOG(DEBUG, "cps update: [%s] family=%u dst_len=%u src_len=%u tos=%u table=%u protocol=%u scope=%u type=%u flags=%x\n",
 		req->nlmsg_type == RTM_NEWROUTE ? "NEW" : "DEL",
 		rm->rtm_family, rm->rtm_dst_len, rm->rtm_src_len,
 		rm->rtm_tos, rm->rtm_table, rm->rtm_protocol,
 		rm->rtm_protocol, rm->rtm_scope, rm->rtm_flags);
 
-	update->type = req->nlmsg_type;
-	update->family = rm->rtm_family;
+	memset(&update, 0, sizeof(update));
+	update.valid = false;
+	update.type = req->nlmsg_type;
+	update.family = rm->rtm_family;
 
 	/* Destination prefix length, e.g., 24 or 32 for IPv4. */
-	update->prefix_len = rm->rtm_dst_len;
+	update.prefix_len = rm->rtm_dst_len;
 
 	/* Default to an invalid index number. */
-	update->oif_index = 0;
+	update.oif_index = 0;
 
 	/* Route origin (routing daemon). */
-	update->rt_proto = rm->rtm_protocol;
+	update.rt_proto = rm->rtm_protocol;
 
 	switch(rm->rtm_family) {
 	case AF_INET:
 		mnl_attr_parse(req, sizeof(*rm), data_ipv4_attr_cb, tb);
-		attr_get(update, rm->rtm_family, tb);
+		attr_get(&update, rm->rtm_family, tb);
 		break;
 	case AF_INET6:
 		mnl_attr_parse(req, sizeof(*rm), data_ipv6_attr_cb, tb);
-		attr_get(update, rm->rtm_family, tb);
+		attr_get(&update, rm->rtm_family, tb);
 		break;
 	default:
 		CPS_LOG(NOTICE, "Unrecognized family in netlink event: %u\n",
 			rm->rtm_family);
 		break;
 	}
 
+	if (update.valid) {
+		if (update.type == RTM_NEWROUTE)
+			new_route(&update, cps_conf);
+		else
+			del_route(&update, cps_conf->gk);
```

Comment:
Let's keep the robust code that there was before and bring back the `else if` that tested for `RTM_DELROUTE`.